### PR TITLE
Fix tab navigation within modal dialog

### LINKED ISSUE #7594: A11y_AzureDataStudio_AddNewController_Keyboard : Loss of keyboard focus occur after the "Cancel" button.
"[Check out Accessibility Insights!](https://accessibilityinsights.io/)- Identify accessibility bugs before check-in and make bug fixing faster and easier."
​
GitHubTags:#A11y_AzureDataStudio;#A11yMAS;#A11yTCS;#WCAG2.4.7;DesktopApp;#Keyboard;#SQL Azure Data Studio;#Win32

**Environment Details:**
Application Name: Azure Data Studio
Version: 1.12.0-insider (user setup)
Commit: 6582debd7388b1be512fc3a9b956d3e54f6115e7
Date: 2019-10-03T08:56:27.454Z
VS Code: 1.38.0
Electron: 4.2.10
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.18362

**Additional Details:**
Tool Name : Keyboard

**Steps to Reproduce:**
1. Install and Launch the application the azure data studio application.
2. Connect to the servers with the required credentials.
3. Navigate to the "SQL Server Big Data Cluster" and expand it.
4. Navigate to the "SQL Server Big Data Cluster controller" and click on it.
5. Click on the "Add New Controller" using mouse and navigate till "Cancel" button.
6. Observe that after the "Cancel" button the keyboard focus is not visible .

**Actual:**
After the "Cancel" button , the keyboard focus is not visible on any control after pressing first tab. Pressing tab again the keyboard focus is visible on background.

**Expected:**
After the "Cancel" button , the keyboard focus should be visible on the "URL" edit box.

**User Impact:**
The keyboard only users will get confused while navigating the screen.

**Attachment for Reference:**
[7594_A11y_AzureDataStudio_AddNewController_Keyboard.pptx](https://github.com/microsoft/azuredatastudio/files/3707224/7594_A11y_AzureDataStudio_AddNewController_Keyboard.pptx)

Does this issue occur when all extensions are disabled?: Yes

### LINKED ISSUE #7798: A11y_AzureDataStudio_AddNewController_Keyboard : The keyboard focus is not visible after "cancel" button.
"[Check out Accessibility Insights!](https://accessibilityinsights.io/)- Identify accessibility bugs before check-in and make bug fixing faster and easier."
​
GitHubTags:#A11y_AzureDataStudio;#A11yMAS;#A11yTCS;#WCAG2.4.7;#DesktopApp;#Keyboard;#SQL Azure Data Studio;#Mac

**Environment Details:**
Application Name: Azure Data Studio
Version: 1.12.0-insider (user setup)
Commit: 6582debd7388b1be512fc3a9b956d3e54f6115e7
Date: 2019-10-03T08:56:27.454Z
VS Code: 1.38.0
Electron: 4.2.10
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: macOS Mojave Version 10.14.6(18G87)

**Additional Details:**
Tool Name : Keyboard

**Steps to Reproduce:**
1. Install and Launch the application the azure data studio application.
2. Connect to the servers with the required credentials.
3. Navigate to the "SQL Server Big Data Cluster" and expand it.
4. Navigate to the "SQL Server Big Data Cluster controller" and click on it.
5. Click on the "Add New Controller" using mouse and navigate till "Cancel" button.
6. Observe that after the "Cancel" button the keyboard focus is not visible .

**Actual:**
After the "Cancel" button , the keyboard focus is not visible on any control after pressing first tab. Pressing tab again the keyboard focus is visible on background.

**Expected:**
After the "Cancel" button , the keyboard focus should be visible on the "URL" edit box.

**User Impact:**
The keyboard only users will get confused while navigating the screen.

**Attachment for Reference:**
[7798_A11y_AzureDataStudio_AddNewController_Keyboard.pptx](https://github.com/microsoft/azuredatastudio/files/3742778/7798_A11y_AzureDataStudio_AddNewController_Keyboard.pptx)

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by Charles-Gagnon**
Fixes #7798
Fixes #7594

This fixes an issue where navigating forwards/backwards from the last/first element in a dialog would escape the dialog. This happened because the logic to find the "first" tabbable element wasn't taking into account the visibility of the elements and so the "first" elements for the dialog were the buttons on the error message div (which are hidden by default). 

Since the tab could never reach that then when you were on the first visible element and tried to tab backwards the logic would just fall back to the default behavior and thus go to other elements behind the dialog.

The fix here is that instead of just showing/hiding elements in the DOM we add/remove them completely - so screen readers and other accessibility tools don't even see them.

### CODE REVIEW

**Comment by anthonydresser** [comment]
Why are the elements that aren't visible just removed from the dom? rather than set to be hidden?

**Comment by Charles-Gagnon** [comment]
Not sure. That would potentially be another way to fix this but I don't really see a huge benefit one way or another - and doing it this way makes it so if other things are hidden they don't affect the tab order either. 

If you feel strongly about doing it that way though I can change it, just seemed like more/riskier work.

File: src/sql/workbench/browser/modal/modal.ts [hunk 1]
```diff
@@ -320,7 +321,7 @@ export abstract class Modal extends Disposable implements IThemable {
 	 * Figures out the first and last elements which the user can tab to in the dialog
 	 */
 	public setFirstLastTabbableElement() {
-		let tabbableElements = this._bodyContainer.querySelectorAll('a[href], area[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled]), [tabindex="0"]');
+		const tabbableElements = this.getBodyContainerTababbleElements();
```
Comment (abist) [inline src/sql/workbench/browser/modal/modal.ts]: nit: the function should be named `getBodyContainerTabbableElements()` instead

**Comment by anthonydresser** [comment]
Its not a huge issue, but it sticks out to me that there might be other problems in this area with how it currently behavior since it doesn't seem as much of a _fix_ as a _patch_

**Comment by Charles-Gagnon** [comment]
@anthonydresser Alright, turned out to be simpler than I thought so ended up going that route

**Pull request merged**